components/Common/SearchBar/index.tsx

```diff
@@ -0,0 +1,337 @@
+import { MdTravelExplore, MdClose } from 'react-icons/md';
+import classNames from 'classnames';
+import { AnimatePresence, motion } from 'framer-motion';
+import Link from 'next/link';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { FormattedMessage } from 'react-intl';
+import styles from './index.module.scss';
+import SectionTitle from '../SectionTitle';
+import { useClickOutside } from '../../../hooks/useClickOutside';
+import useKeyPress from '../../../hooks/useKeypress';
+import type { FC } from 'react';
+import type { SearchFunction, SearchResult } from '../../../types';
+
+type SearchBarProps = {
+  search?: SearchFunction;
+  setup?: () => Promise<void>;
+  footer?: FC | (() => JSX.Element);
+};
+
+const containerTransition = { type: 'spring', damping: 22, stiffness: 150 };
+
+const containerVariants = {
+  expanded: {
+    width: 'auto',
+    height: 'auto',
+  },
+  collapsed: {
+    width: 0,
+    height: '3em',
+  },
+};
+
+const SearchBar: FC<SearchBarProps> = ({
+  setup,
+  search,
+  footer: Footer,
+}: SearchBarProps) => {
+  const [isLoading, setIsLoading] = useState(true);
+  const searchInputRef = useRef<HTMLInputElement | null>(null);
+  const listRef = useRef<HTMLUListElement | null>(null);
+  const activeResultRef = useRef<number>(-1);
+
+  const [isExpanded, setExpanded] = useState(false);
+  const [query, setQuery] = useState('');
+  const [results, setResults] = useState<SearchResult[]>([]);
+  const isEmpty = !results || results.length === 0;
```

Comment:
I feel like this Component is doing to many things. The interactions  between a lot of the state objects is making it complicated. Can we try to use maybe `useReducer` to see if we can clean things up here?

Reply:
I'm -1 on this.

`query` and `results` are basically changed at the same time.
`expanded` is triggered independently.
The only other state is `isLoading` which is only changed once.

I feel adding adding a reducer would be slightly overengineered.